pyo3-macros-backend/src/frompyobject.rs

```diff
@@ -233,19 +241,33 @@ impl<'a> Container<'a> {
         }
     }
 
-    fn build_tuple_struct(&self, len: usize) -> TokenStream {
+    fn build_tuple_struct(&self, tups: &[FieldPyO3Attributes]) -> TokenStream {
         let self_ty = &self.path;
         let mut fields: Punctuated<TokenStream, syn::Token![,]> = Punctuated::new();
-        for i in 0..len {
-            let error_msg = format!("failed to extract field {}.{}", quote!(#self_ty), i);
-            fields.push(quote!(
-                s.get_item(#i).and_then(_pyo3::types::PyAny::extract).map_err(|inner| {
-                let py = _pyo3::PyNativeType::py(obj);
-                let new_err = _pyo3::exceptions::PyTypeError::new_err(#error_msg);
-                new_err.set_cause(py, ::std::option::Option::Some(inner));
-                new_err
-                })?));
+        for (index, attrs) in tups.iter().enumerate() {
+            let error_msg = format!("failed to extract field {}.{}", quote!(#self_ty), index);
+
+            let extractor = match &attrs.from_py_with {
+                None => quote!(
```

Comment:
The indentation here looks off, i.e. the block inside `map_err` should be indented one to the right?

Reply:
You are completely right, I've also extracted the duplication of code.